File: cpp/structure/tree/TrieTree.cpp
#include <iostream>
#include <vector>
using namespace std;

class TrieTree {
public:
    TrieTree() : root(new TreeNode()) {}

    bool isEmpty() {
        for (int i = 0; i < 26; ++i) {
            if (root->next[i]) {
                return false;
            }
        }
        return true;
    }

    void add(string text) {
        TreeNode *current = root;
        for (char value : text) {
            int index = value - 'a';
            if (!current->next[index]) {
                current->next[index] = new TreeNode();
                ++current->depth;
            }
            current = current->next[index];
        }
        current->ended = true;
    }

    void erase(string text) {
        int index = text[0] - 'a';
        root->next[index] = eraseNode(root->next[index], text, text.size(), 1);
    }

    bool contains(string text) {
        TreeNode *current = root;
        for (char value : text) {
            int index = value - 'a';
            TreeNode *next = current->next[index];
            if (!next) {
                return false;
            }
            current = next;
        }
        return current->ended;
    }

    void clear() {
        for (int i = 0; i < 26; ++i) {
            root->next[i] = nullptr;
        }
    }

private:
    struct TreeNode {
        TreeNode **next;
        int depth;
        bool ended;

        TreeNode(): depth(0), ended(false) {
            next = new TreeNode *[26];
            for (int i = 0; i < 26; ++i) {
                next[i] = nullptr;
            }
        }
    };

    TreeNode *root;

    TreeNode *eraseNode(TreeNode *current, string text, int depth, int pos) {
        if (!current) {
            return nullptr;
        }

        if (pos >= depth) {
            if (!current->depth) {
                return nullptr;
            }
            current->ended = false;
            return current;
        }

        int index = text[pos] - 'a';
        current->next[index] = eraseNode(current->next[index], text, depth, pos + 1);
        if (current->next[index]) {
            return current;
        }

        if (current->depth) {
            --current->depth;
        }
        if (!current->depth && !current->ended) {
            return nullptr;
        }
        return current;
    }
};

File: py/structure/tree/TrieTree.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class TrieTree:
    def __init__(self):
        self.__root = self.__TreeNode()

    def isEmpty(self) -> bool:
        for i in range(26):
            if self.__root.next[i]:
                return False
        return True

    def add(self, text: str) -> None:
        current = self.__root
        for value in text:
            index = ord(value) - ord('a')
            if not current.next[index]:
                current.next[index] = self.__TreeNode()
                current.depth += 1
            current = current.next[index]
        current.ended = True

    def erase(self, text: str) -> None:
        index = ord(text[0]) - ord('a')
        self.__root.next[index] = self.__eraseNode(self.__root.next[index], text, len(text), 1)

    def contains(self, text: int) -> bool:
        current = self.__root
        for value in text:
            index = ord(value) - ord('a')
            next = current.next[index]
            if not next:
                return False
            current = next
        return current.ended

    def clear(self) -> None:
        for i in range(26):
            self.__root.next[i] = None

    def __eraseNode(self, current: '__TreeNode', text: str, depth: int, pos: int) -> '__TreeNode':
        if not current:
            return None

        if pos >= depth:
            if not current.depth:
                return None
            current.ended = False
            return current

        index = ord(text[pos]) - ord('a')
        current.next[index] = self.__eraseNode(current.next[index], text, depth, pos + 1)
        if current.next[index]:
            return current

        if current.depth:
            current.depth -= 1
        if not current.depth and not current.ended:
            return None
        return current

    class __TreeNode:
        def __init__(self):
            self.next, self.depth, self.ended = [None for _ in range(26)], 0, False
